http/basic/src/main/java/io/cloudevents/http/HttpMessageFactory.java

```diff
@@ -0,0 +1,166 @@
+     *
+     *     }
+     * };
+     * byte[] body;
+     * String contentLength = httpServletRequest.getHeader("Content-Length");
+     * if (contentLength != null) {
+     *     body = IOUtils.toByteArray(httpServletRequest.getInputStream(), Integer.parseInt(contentLength));
+     * } else {
+     *     body = IOUtils.toByteArray(httpServletRequest.getInputStream());
+     * }
+     * HttpMessageFactory.createReader(forEachHeader, body);
+     * }
+     * </pre>
+     * @param forEachHeader http headers visitor function
+     * @param body          nullable buffer of the body
+     * @return a message reader implementation with potentially an unknown encoding
+     * @throws IllegalArgumentException If, in case of binary mode, the spec version is invalid
+     */
+    public static MessageReader createReader(Consumer<BiConsumer<String, String>> forEachHeader, byte[] body) {
+        final String[] hdrs = {null, null};
+
+        forEachHeader.accept((k, v) -> {
+            if (CONTENT_TYPE.equalsIgnoreCase(k)) {
+                hdrs[0] = v;
+            } else if (CloudEventsHeaders.SPEC_VERSION.equalsIgnoreCase(k)) {
+                hdrs[1] = v;
+            }
+        });
+
+        final String contentType = hdrs[0];
+        final String specversion = hdrs[1];
+
+        return MessageUtils.parseStructuredOrBinaryMessage(
+            () -> contentType,
+            format -> new GenericStructuredMessageReader(format, body),
+            () -> specversion,
+            sv -> new HttpMessageReader(sv, forEachHeader, body),
+            UnknownEncodingMessageReader::new
+        );
+    }
+
+    /**
+     * Creates a new {@link MessageReader} that can read both structured and binary messages from a HTTP response (client) or request (server).
+     *
+     * <pre>
+     * This overload is equivalent to calling:
+     * {@code HttpMessageFactory.createReader(headers::forEach, body);}
+     * </pre>
+     * @param headers http headers as map
+     * @param body    nullable buffer of the body
+     * @return a message reader implementation with potentially an unknown encoding
+     * @throws IllegalArgumentException If, in case of binary mode, the spec version is invalid
+     */
+    public static MessageReader createReader(Map<String, String> headers, byte[] body) {
+        return createReader(headers::forEach, body);
+    }
+
+    /**
+     * Creates a new {@link MessageReader} that can read both structured and binary messages from a HTTP response (client) or request (server).
+     *
+     * @param headers http headers as multimap
+     * @param body    nullable buffer of the body
+     * @return a message reader implementation with potentially an unknown encoding
+     * @throws IllegalArgumentException If, in case of binary mode, the spec version is invalid
+     */
+    public static MessageReader createReaderFromMultiMap(Map<String, List<String>> headers, byte[] body) {
```

Comment:
I think the spelling Multimap rather than MultiMap is more usual.

Reply:
I prefer MultiMap because to me it's like Multimap, but it's a matter of taste

Reply:
A `TreeMap` is just a kind of `Map`, but a `Multimap` is a different sort of thing. As words we would write "tree map" but "multimap". https://en.wikipedia.org/wiki/Multimap